apps/sms/views/shared/js/activity_handler.js

```diff
@@ -21,22 +23,27 @@ const ActivityDataType = {
 };
 
 var ActivityHandler = {
-  // Will hold current activity object
-  _activity: null,
+  _activityClient: null,
 
   init: function() {
     if (!window.navigator.mozSetMessageHandler) {
       return;
     }
 
-    // A mapping of MozActivity names to their associated event handler
-    window.navigator.mozSetMessageHandler('activity',
-      this._onActivity.bind(this, {
-        'new': this._onNewActivity,
-        'share': this._onShareActivity
-      })
+    this._activityClient = new ActivityClient();
+
+    this._activityClient.on(
+      'new-activity-request', this._onNewActivity.bind(this)
+    );
+
+    this._activityClient.on(
+      'share-activity-request', this._onShareActivity.bind(this)
     );
 
+    // Currently both client and service for activity resides in the same
+    // context, so we instantiate both.
+    (new ActivityService());
```

Comment:
Let me describe what problem we have currently:

Case №1:
1. When we run Messages app we create service instance with a name "A";
2. Then we create client instance bound to service "A";

Everything works fine in this case; Hope it's by intention :)

Case №2:
1. When we run Messages app we create service #1 with a name "A";
2. Then we create client#1 instance bound to service "A";
3. Then we run one more instance via activity system message and create service#2 "A" again;
4. Then one more client#2 for service "A".

I haven't debug threads.js yet, but that's how I think it works currently:
Service#2 is ignored, like when we create service#2, we see that we already have such service (via BroadcastChannel?) and when client#2 calls service method, then service#1 processes this message only.

If my guess is correct I think we can solve it in the following ways:
* If app is run not as activity (mozHasPendingMessage?) we don't create activity service - everthing should be fine here;
* ActivityService and ActvivityClient can use "context-bound" service name (like threads.service('activity-service-' + App.Id)) where App.Id maybe just Date.now() called on app startup;
* We can improve threads.js to say that service we register is in-process only, so clients from another processes can't access it.

I'll debug and check if my guess isn't stupid :)